engines/agi/loader_gal_a2.cpp

```diff
@@ -0,0 +1,292 @@
+}
+
+/**
+ * Reads a directory entry.
+ */
+bool GalLoader_A2::readDirectoryEntry(Common::SeekableReadStream &stream, AgiDir &dirEntry) {
+	// GAL A2 DIR format:
+	// track   8 bits
+	// disk    4 bits (0 for all disks, else 1-3)
+	// sector  4 bits
+	// offset  8 bits
+	byte b0 = stream.readByte();
+	byte b1 = stream.readByte();
+	byte b2 = stream.readByte();
+
+	byte disk = b1 >> 4;
+	byte sector = b1 & 0x0f;
+	uint32 position = A2_DISK_POSITION(b0, sector, b2);
+
+	// use the first disk for resources that are on all disks
+	if (disk > 0) {
+		disk--;
+	}
+
+	// validate entry
+	if (!(disk <= 2 && position < A2_DISK_SIZE)) {
+		return false;
+	}
+
+	dirEntry.volume = disk;
+	dirEntry.offset = position;
+	return true;
+}
+
+/**
+ * Tests if a disk contains all of the expected logic resources.
+ */
+bool GalLoader_A2::validateDisk(Common::SeekableReadStream &disk, byte diskIndex, AgiDir *logicDir) {
+	for (int i = 0; i < GAL_A2_LOGIC_COUNT; i++) {
+		// only validate logics on this disk
+		if (logicDir[i].volume != diskIndex) {
+			continue;
+		}
+
+		// skip logic 64 because it contains incorrect header values
+		if (i == 64) {
+			continue;
+		}
+
+		// GAL A2 resources begin with a length before the actual resource.
+		// Logic resources begin with a header of four lengths. If a logic entry
+		// is valid, then the A2 length will match the calculated logic length.
+		disk.seek(logicDir[i].offset);
+		uint16 resourceLength = disk.readUint16LE();
+		uint32 logicLength = 8;
+		for (int j = 0; j < 4; j++) {
+			logicLength += disk.readUint16LE();
+		}
+		if (disk.eos() ||
+			resourceLength != logicLength ||
+			!(logicDir[i].offset + 2 + resourceLength <= A2_DISK_SIZE)) {
+			return false;
+		}
+	}
+	return true;
+}
+
+/**
+ * Load logic, pic, and view directories. KQ1-A2 has no sound resources.
+ */
+int GalLoader_A2::loadDirs() {
+	// if init didn't find disks then fail
+	if (_disks.empty()) {
+		return errFilesNotFound;
+	}
+	for (uint d = 0; d < _disks.size(); d++) {
+		if (_disks[d] == nullptr) {
+			warning("AgiLoader_A2: disk %d not found", d);
+			return errFilesNotFound;
+		}
+	}
+
+	// directories are on disk one
+	Common::SeekableReadStream &disk = *_disks[0];
+
+	bool success = true;
+	success &= loadDir(_vm->_game.dirLogic, disk, GAL_A2_LOGDIR_POSITION,  GAL_A2_LOGIC_COUNT);
+	success &= loadDir(_vm->_game.dirPic,   disk, GAL_A2_PICDIR_POSITION,  GAL_A2_PICTURE_COUNT);
+	success &= loadDir(_vm->_game.dirView,  disk, GAL_A2_VIEWDIR_POSITION, GAL_A2_VIEW_COUNT);
+	return success ? errOK : errBadResource;
+}
+
+/**
+ * Loads a resource directory.
+ */
+bool GalLoader_A2::loadDir(AgiDir *dir, Common::SeekableReadStream &disk, uint32 dirOffset, uint32 dirCount) {
+	disk.seek(dirOffset);
+	for (uint32 i = 0; i < dirCount; i++) {
+		// skip picture 0 and 81, are empty and 81's entry contains junk bytes
+		// that would fail validation.
+		if ((i == 0 || i == 81) && dirOffset == GAL_A2_PICDIR_POSITION) {
+			disk.skip(3);
+			continue;
+		}
```

Comment:
Is this considered a workaround? How does the original handle it?

Reply:
Clarified in the comments; this is up-front parsing/validation/disk-detection that the original didn't do.